platform/src/server/http/http_server.ts

```diff
@@ -1,95 +1,243 @@
-    // '((request: IncomingMessage, response: ServerResponse) => void) | undefined'
-    // ```
-    const app = this.app as any;
-
-    if (!config.ssl.enabled) {
-      return http.createServer(app);
-    }
+    const options: ServerOptions = {
+      host: config.host,
+      port: config.port,
+      routes: {
+        cors: config.cors,
+        payload: {
+          maxBytes: config.maxPayload.getValueInBytes(),
+        },
+        validate: {
+          options: {
+            abortEarly: false,
+          },
+        },
+      },
+      state: {
+        strictHeader: false,
+      },
+    };
 
-    // TODO: add support for `secureOptions`.
-    return https.createServer(
-      {
-        key: readFileSync(config.ssl.key!),
-        cert: readFileSync(config.ssl.certificate!),
+    const ssl = config.ssl;
+    if (ssl.enabled) {
+      const tlsOptions: TLSOptions = {
         ca:
           config.ssl.certificateAuthorities &&
           config.ssl.certificateAuthorities.map(caFilePath =>
             readFileSync(caFilePath)
           ),
-        passphrase: config.ssl.keyPassphrase,
+
+        key: readFileSync(ssl.key!),
+        cert: readFileSync(ssl.certificate!),
+        passphrase: ssl.keyPassphrase,
+
         ciphers: config.ssl.cipherSuites.join(':'),
         // We use the server's cipher order rather than the client's to prevent the BEAST attack.
         honorCipherOrder: true,
-      },
-      app
+        secureOptions: ssl.getSecureOptions(),
+      };
+
+      // TODO: Hapi types have a typo in `tls` property type definition: `https.RequestOptions` is used instead of
+      // `https.ServerOptions`, and `honorCipherOrder` isn't presented in `https.RequestOptions`.
+      options.tls = tlsOptions as any;
+    }
+
+    return new Server(options);
+  }
+
+  private _setupBasePathRewrite(server: Server, config: HttpConfig) {
+    if (config.basePath === undefined || !config.rewriteBasePath) {
+      return;
+    }
+
+    const basePath = config.basePath;
+    server.ext(
+      'onRequest',
+      (request: Request, responseToolkit: ResponseToolkit) => {
+        const newURL = modifyUrl(request.url.href!, urlParts => {
+          if (
+            urlParts.pathname != null &&
+            urlParts.pathname.startsWith(basePath)
+          ) {
+            urlParts.pathname = urlParts.pathname.replace(basePath, '') || '/';
+          } else {
+            return {};
+          }
+        });
+
+        if (!newURL) {
+          return responseToolkit
+            .response('Not Found')
+            .code(404)
+            .takeover();
+        }
+
+        request.setUrl(newURL);
+        // We should update raw request as well since it can be proxied to the old platform
+        // where base path isn't expected.
+        request.raw.req.url = request.url.href;
+
+        return responseToolkit.continue;
+      }
     );
   }
+
+  private async _setupRedirectServer(config: HttpConfig) {
```

Comment:
__note:__ now when added it here, I'm thinking maybe I should have `http_redirect_server.ts` instead that is managed by `http_service` as it seems to be independent enough from `http_server` itself?

Reply:
I agree.

Reply:
I'll handle this in the follow-up then.